ydb/library/yql/dq/comp_nodes/dq_join_common.h

```diff
@@ -339,8 +315,443 @@ template <typename Source> class TJoinPackedTuples {
+            MKQL_ENSURE(Data_.Probe, "trying to clone Source");
+            return std::move(*Data_.Probe);
+        }
+
+      private:
+        TSides<std::optional<Source>> Data_;
+    };
+
+    THybridHashJoin(TSides<Source> sources, NUdf::TLoggerPtr logger, TString componentName,
+                    TSides<const NPackedTuple::TTupleLayout*> layouts, TComputationContext& ctx)
+        : Logger_(logger, logger->RegisterComponent(componentName))
+        , Layouts_(layouts)
+        , Spiller_(ctx.SpillerFactory->CreateSpiller())
+        , Sources_(std::move(sources))
+    {}
+
+    struct Finish {};
+
+    TPackResult Flatten(TMKQLVector<TPackResult> tuples) {
+        return Layouts_.Build->Flatten(tuples);
+    }
+
+    EFetchResult WaitWhileSpilling() {
+        return EFetchResult::Yield;
+    }
+
+    TPackResult GetPage(TFuturePage&& future) {
+        std::optional<NYql::TChunkedBuffer> buff = ExtractReadyFuture(std::move(future));
+        MKQL_ENSURE(buff.has_value(), "corrupted extract key?");
+        return Parse(std::move(*buff));
+    }
+
+    EFetchResult MatchRows([[maybe_unused]] TComputationContext& ctx,
+                           std::invocable<TSides<TSingleTuple>> auto consumePairOfTuples) {
+        auto fetchSpillingCondition = [](TPackResult& res) {
+            return [&] {
+                if (!TlsAllocState->GetMaximumLimitValueReached()) {
+                    return false;
+                }
+                return ui64(FreeMemory()) < std::max(ui64(res.AllocatedBytes()), 200_KB);
+            };
+        };
+        if (std::get_if<Init>(&State_)) {
+            State_ = FetchingBuild{*this};
+        } else if (auto* s = std::get_if<FetchingBuild>(&State_)) {
+            FetchingBuild& state = *s;
+            if (!state.Pack.has_value()) {
+                FetchResult<TPackResult> var = state.Build.FetchRow();
+                NYql::NUdf::EFetchStatus status = AsStatus(var);
+                if (status == NYql::NUdf::EFetchStatus::Yield) {
+                    return EFetchResult::Yield;
+                } else if (status == NYql::NUdf::EFetchStatus::Ok) {
+                    state.Pack = std::move(GetPayload(var));
+                } else {
+                    MKQL_ENSURE(status == NYql::NUdf::EFetchStatus::Finish, "unhandled status");
+                    MKQL_ENSURE(state.Build.Finished(), "sanity check");
+                    State_ = BuildingInMemoryTable{*this, std::move(state.Spiller)};
+                }
+            } else {
+                ESpillResult res = state.Spiller.SpillWhile(fetchSpillingCondition(*state.Pack));
+                switch (res) {
+                case Spilling:
+                    return WaitWhileSpilling();
+                case FinishedSpilling:
+                    break;
+                case DontHavePages:
+                    MKQL_ENSURE(false,
+                                "spilling in smaller pages is not implemented"); // we can not spill much and do not
```

Comment:
Does it mean, that we spilled all the pages from all the buckets and still don't have enough memory? And what does "spilling in smaller pages" mean here? Flushing pages that are not fully filled?